### PR TITLE
Hardcoded Gas Limit

### PR DESCRIPTION
**Pull request by PhABC**
.

### CODE REVIEW

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:2]:
```suggestion
eip: 3382
```

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:10]:
```suggestion
```

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:15]: Since this EIP does not require 1559, I recommend using non-1559 terminology "gas limit".

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:23]: I'll need to check later if the standard easy of referring to POW and POS is hyphenated or not.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:26]: If so, I recommend removing this and requiring 1559 in the header.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:30]: A nit, it is probably easier to keep the gas limit in the block and ignore than remove. But since we're adding BASEFEE already, the point may be moot.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:32]: This is already implied by 1559.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:36]: I don't think this needs to be specified as it is implied.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:52]: Will also break on-chains contracts that verify headers.

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:14]: Good point. I'll leave a note about EIP-1559 renaming

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:51]: Fair point, though wouldn't any change to the block structure result to this?

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:31]: Is it? I believe gas_target is still part of block headers with EIP-1559.

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:29]: Indeed, I believe EIP-1559 already forces a change to the data structure to be hashed for block headers.

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:35]: I'll leave it to be explicit about the `<` instead of `<=`, which could cause a fork.

File: EIPS/eip-hard-gas-target.md
Comment (MicahZoltu) [reply EIPS/eip-hard-gas-target.md]: 1559 adds an item on the end, so if you have code that hashes the block and/or plucks items out of it by offset then 1559 won't break anything.  However, if we pull gas limit out (rather than just ignore it) any code that plucks data *after* the gas limit will break.

File: EIPS/eip-hard-gas-target.md
Comment (MicahZoltu) [reply EIPS/eip-hard-gas-target.md]: No, see comment above.

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md]: Fair, then for simplicity I'll remove this aspect from the EIP. Even though it would become mainly a cosmetic field, we probably need to add a check in the consensus that the gasLimit provided in the header is equal to the hardcoded gas limit, in case people use the former. Would you both agree?

**Review by KaycoinsInc — APPROVED** [review]

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [reply EIPS/eip-hard-gas-target.md:27]: I'm referring to the consensus section here. If the gas limit is required to be 12.5mm per consensus then the gas used will be appropriately constrained pre and post-1559.

Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:27]: Ah, I thought your original comment was referring to `As of `FORK_BLOCK_NUMBER`, remove `gas_target` field from block headers.`

However, EIP-1559 does not imply the hardcoded gasLimit is part of the consensus. Current EIP-1559 spec uses the gas_target from a block's header, not this proposed hardcoded constant. As per previous comments you and Micah made, the gas_target provided in headers will remain, so it's important that the consensus algorithm implementations are changed to using the hard-coded gasLimit, not the header's gas limit.

Comment (lightclient) [reply EIPS/eip-hard-gas-target.md:27]: Ah yes, if the limit was removed from the header it would need to be specified. Since the limit will stay in the header, why not a consensus rule that requires `block.gas_limit == 12_500_000`? The rules for EIP-1559 should stay the same in that case?

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [reply EIPS/eip-hard-gas-target.md:31]: I was actually referring to the EVM section here. I realize now that if the header *didn't* have `gas_limit` specified, this would be needed. However, if the header does include the `gas_limit` the result of `GASLIMIT` should be implied.

Comment (lightclient) [reply EIPS/eip-hard-gas-target.md:31]: > I'll leave it to be explicit about the < instead of <=, which could cause a fork if not implemented the same.

What is the rationale for `<` vs. `<=`? I believe today `gas_used` must be less than or equal to `gas_limit`, seems like it is better to maintain that unless there is a particular reason to change?

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md:27]: Good point, i'll remove the redundant stuff.

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md]: Searched and found a few EIPs that use these terms the same way, e.g. https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1011.md and https://eips.ethereum.org/EIPS/eip-1057#motivation while others do used an hyphenated version like https://eips.ethereum.org/EIPS/eip-3368 and  https://eips.ethereum.org/EIPS/eip-1057's title.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:5]: Please add a discussion link when you're ready for this to be merged.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [inline EIPS/eip-hard-gas-target.md:48]:
```suggestion
```

**Review by lightclient — APPROVED** [review]
EIP is ready to be merged once the discussion link is filled in.

File: EIPS/eip-hard-gas-target.md
Comment (lightclient) [reply EIPS/eip-hard-gas-target.md]: Doesn't appear to be a strong consensus on one version or the other. Please use your preferred. If certain style is decided on the future, all EIPs will be updated accordingly.

```console
$ grep -ril "proof of work" EIPS | wc -l
9
$ grep -ril "proof-of-work" EIPS | wc -l
13
$ grep -ril "proof of stake" EIPS | wc -l
5
$ grep -ril "proof-of-stake" EIPS | wc -l
8
```

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md]: Added

File: EIPS/eip-hard-gas-target.md
Comment (PhABC) [reply EIPS/eip-hard-gas-target.md]: Thanks for doing the search, looks pretty evenly distributed!

File: EIPS/eip-3382.md
Comment (MicahZoltu) [inline EIPS/eip-3382.md:14]: Links to external EIPs should be relative (`[EIP-1559](./eip-1559.md)`).  That being said, this is not really a detail relevant to the simple summary.  The simple summary should be just a short one liner that says what the EIP is.
```suggestion
Hardcode the block gas target to `12,500,000` gas per block.
```

File: EIPS/eip-3382.md
Comment (MicahZoltu) [inline EIPS/eip-3382.md:18]: Abstract should be a human readable and terse version of the specification.  The first sentence you have talks about history, which isn't appropriate for the specification/abstract.  The second sentence is a tad less clear than I think would be ideal.

```suggestion
Updates the block validation rules such that a block is invalid if the `gas_target` header field is not equal to 12,500,000.
```

File: EIPS/eip-3382.md
Comment (MicahZoltu) [inline EIPS/eip-3382.md:26]: Write this either under the assumption of EIP-1559 or not.  We can always change the wording one way or the other at the last minute should plans change since that would be a non-normative change.  I recommend writing this under the assumption of EIP-1559, since that seems very likely to land in London.
```suggestion
```

File: EIPS/eip-3382.md
Comment (MicahZoltu) [inline EIPS/eip-3382.md:29]: Instead of `header.gasLimit`, I recommend referring to it by its position in the header tuple.  `gasLimit` is what most people call it, but from a consensus point of view it is the `n`th (I forget the number) item in the block header array.

Comment (MicahZoltu) [reply EIPS/eip-3382.md:29]: See note above about using gas **target** throughout this document, including here.

**Review by MicahZoltu — APPROVED** [review]
Some changes requested, but nothing blocking merge as draft.

File: EIPS/eip-3382.md
Comment (PhABC) [reply EIPS/eip-3382.md:26]: I initially used gasTarget across the entire EIP, however as @lightclient pointed out, these two EIPs are independant so it might be clearer if we use the current term.

File: EIPS/eip-3382.md
Comment (PhABC) [reply EIPS/eip-3382.md:29]: See my reply above, but will add tuple number

File: EIPS/eip-3382.md
Comment (MicahZoltu) [reply EIPS/eip-3382.md:26]: I agree with @lightclient that this EIP is independent of 1559 and should be written as such.  However, you should write a single EIP as an assertion of a change, without conditionals, which means you have to choose one wording or the other.  In this case, I think EIP-1559 is more likely to be included in London than not, so the "safer" bet is to use "gas target" wording as that minimizes the chances that you will need to update the EIP later should the environment change.

**Pull request merged**
